File: python/src/azoth/eos/reference/ph_flash.py
"""``eos.ph_flash`` - the temperature a mixture reaches at a given pressure and enthalpy.

Spec: ``specs/models/eos/ph_flash.toml``

An outer bisection on temperature, over the enthalpy assembled from two things this
library already has: the phase split at a trial temperature, from
:mod:`azoth.eos.reference.pt_flash`, and each phase's enthalpy, from
``eos.molar_enthalpy_entropy``. The enthalpy rises monotonically with temperature at a
fixed pressure, which is what makes the bisection well posed.

The bracket, the loop and the phase branch live in
:mod:`azoth.eos.reference._flash_property`, shared with ``eos.ps_flash``.
"""

from __future__ import annotations

from typing import Any

from azoth import _models_gen
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import PhFlashResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.mixture import Mixture
from azoth.eos.reference._flash_property import Property, distinct_warnings, property_at
from azoth.eos.reference._flash_property import solve_temperature as _solve
from azoth.eos.reference.molar_enthalpy_entropy import IdealGasModel

MODEL_ID = "eos.ph_flash"

#: Which property this model inverts. The one thing it does not share with `ps_flash`.
WHICH: Property = "h"


def enthalpy_at(
    mixture: Mixture,
    ideal_gas: IdealGasModel,
    t_si: float,
    p_si: float,
    z: list[float],
) -> tuple[float, dict[str, Any]]:
    """The molar enthalpy of a mixture at a temperature and pressure, and its split.

    The composition the flash settles on is the equilibrium one, so this is the enthalpy
    of the *feed* at that state - which is what makes it comparable with a duty a caller
    supplied.
    """
    return property_at(mixture, ideal_gas, t_si, p_si, z, WHICH)


def solve_temperature(
    mixture: Mixture,
    ideal_gas: IdealGasModel,
    p_si: float,
    h_si: float,
    z: list[float],
    algorithm: dict[str, Any],
) -> dict[str, Any]:
    """Invert ``H(T, P) = h_si`` for ``T`` by bracketing, then bisecting."""
    return _solve(mixture, ideal_gas, p_si, h_si, z, WHICH, algorithm)


def ph_flash(
    mixture: Mixture,
    ideal_gas: IdealGasModel,
    P: Q,
    H: Q,
    z: list[float],
) -> PhFlashResult:
    """The temperature at which a mixture has a given molar enthalpy at a pressure.

    ``H`` is a *difference* from the datum ``ideal_gas`` carries, not an absolute
    quantity: two calls with different reference values are not comparable, and their
    difference is a plausible number rather than an error.

    Raises:
        InvalidInputError: if the composition or any ideal-gas vector is the wrong
            length, or if ``z`` is not a composition.
        OutOfRangeError: if ``P`` is not positive, or a range check on the answer fails.
        SolverNotConvergedError: if no temperature on the bracket covers the requested
            enthalpy, or if the bisection reaches its cap.
    """
    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    p_si = input_to_si(spec, "P", P)
    h_si = input_to_si(spec, "H", H)
    apply_checks(checks.on_input, {"P": p_si, "H": h_si}.get, warnings)

    solved = solve_temperature(mixture, ideal_gas, p_si, h_si, list(z), spec["algorithm"])
    warnings.extend(solved["warnings"])
    flash = solved["state"]["flash"]

    apply_checks(
        checks.derived,
        lambda name: solved["T"] if name == "T" else None,
        warnings,
    )

    return PhFlashResult(
        T=from_si(solved["T"], "K"),
        beta=flash.beta,
        x=tuple(flash.x),
        y=tuple(flash.y),
        k=tuple(flash.k),
        phase=flash.phase,
        z_liquid=flash.z_liquid,
        z_vapour=flash.z_vapour,
        iterations=solved["iterations"],
        residual=solved["residual"],
        warnings=distinct_warnings(warnings),
    )

File: crates/azoth-eos/src/ph_flash.rs
//! `eos.ph_flash` - the temperature a mixture reaches at a given pressure and enthalpy.
//!
//! Spec: `specs/models/eos/ph_flash.toml`
//!
//! A damped quasi-Newton on temperature, in `1/T`, over the enthalpy assembled from two
//! things this crate already has: the phase split at a trial temperature, from
//! [`crate::pt_flash`], and each phase's enthalpy, from
//! [`crate::molar_enthalpy_entropy`]. Upstream's `PHflash.solveQ`, which is the scheme
//! NeqSim runs by default - see the spec's assumption on the second-order alternative it
//! carries.
//!
//! The damping, the step clamp, the trial-temperature recovery and the phase branch live
//! in [`crate::flash_property`], shared with `eos.ps_flash`.

use azoth_core::units::{MolarEnergy, Pressure, ThermodynamicTemperature, kelvins};
use azoth_core::{Result, apply_checks};

use crate::algorithm_of;
use crate::flash_property::{Property, property_at, solve_temperature};
use crate::mixture::Mixture;
use crate::model_gen;
use crate::molar_enthalpy_entropy::IdealGasModel;
use crate::results::{PhFlashResult, PtFlashResult};

/// The molar enthalpy of a mixture at a temperature and pressure, and its split.
///
/// The composition the flash settles on is the equilibrium one, so this is the enthalpy
/// of the *feed* at that state - which is what makes it comparable with a duty a caller
/// supplied.
///
/// # Errors
/// Propagates whatever [`crate::pt_flash`] and [`crate::molar_enthalpy_entropy`] raise.
pub fn enthalpy_at(
    mixture: &Mixture,
    ideal_gas: &IdealGasModel,
    t: ThermodynamicTemperature,
    p: Pressure,
    z: &[f64],
) -> Result<(f64, PtFlashResult)> {
    property_at(mixture, ideal_gas, t, p, z, Property::Enthalpy)
}

/// The temperature at which a mixture has a given molar enthalpy at a pressure.
///
/// `h` is a *difference* from the datum `ideal_gas` carries, not an absolute quantity:
/// two calls with different reference values are not comparable, and their difference is
/// a plausible number rather than an error.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `p` is not positive, or a range check on the answer
///   fails.
/// * [`azoth_core::AzothError::InvalidInput`] if the spec declares no starting temperature,
///   which would be a generator bug rather than a caller's.
/// * [`azoth_core::AzothError::SolverNotConverged`] if the iteration reaches its cap.
///
/// # Example
/// ```
/// use azoth_core::units::{joules_per_mole, pascals};
/// use azoth_eos::{databank, ph_flash};
///
/// let (mixture, ideal_gas) = databank::mixture_of(&["methane", "n-butane"], None)
///     .expect("the pair resolves");
/// let r = ph_flash::ph_flash(
///     &mixture,
///     &ideal_gas,
///     pascals(2_000_000.0),
///     joules_per_mole(-5121.329517317879),
///     &[0.6, 0.4],
/// )?;
/// assert!((r.temperature.value - 300.0).abs() < 1.0e-6);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
pub fn ph_flash(
    mixture: &Mixture,
    ideal_gas: &IdealGasModel,
    p: Pressure,
    h: MolarEnergy,
    z: &[f64],
) -> Result<PhFlashResult> {
    let spec = &model_gen::PH_FLASH_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "P" => Some(p.value),
            "H" => Some(h.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let algorithm = algorithm_of(spec)?;
    let solved = solve_temperature(
        mixture,
        ideal_gas,
        p,
        h.value,
        z,
        Property::Enthalpy,
        algorithm,
    )?;

    apply_checks(
        spec.derived_checks(),
        |quantity| (quantity == "T").then_some(solved.temperature),
        &mut warnings,
    )?;

    Ok(PhFlashResult {
        temperature: kelvins(solved.temperature),
        beta: solved.flash.beta,
        x: solved.flash.x,
        y: solved.flash.y,
        k: solved.flash.k,
        phase: solved.flash.phase,
        z_liquid: solved.flash.z_liquid,
        z_vapour: solved.flash.z_vapour,
        iterations: solved.iterations,
        residual: solved.residual,
        warnings: solved.warnings,
    })
}
